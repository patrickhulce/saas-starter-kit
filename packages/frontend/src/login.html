<!DOCTYPE html>
<html>

<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>THE_PRODUCT_DISPLAY_NAME Login</title>
</head>

<body>
  <h1 class="app-name">THE_PRODUCT_DISPLAY_NAME</h1>
  <div id="app-root"></div>
</body>

</html>
