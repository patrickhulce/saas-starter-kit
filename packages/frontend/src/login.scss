:global {
  html {
    height: 100%;
  }

  body {
    background: #0771b8;
    color: rgba(0, 0, 0, 0.87);
    font-family: Roboto, Helvetica, Arial, sans-serif;
    height: 100%;
    margin: 0;
    padding: 0;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;
  }

  h1.app-name {
    color: white;
    font-size: 48px;
    font-weight: bold;
    text-shadow: 0 1px 5px #444;
    margin-bottom: 40px;
  }

  #firebaseui-auth-container {
    padding: 15px;
    border-radius: 5px;
    background: white;
    box-shadow: 0 5px 15px #444;
    display: none;
  }

  #loader-container {
    height: 250px;
  }

  .loader,
  .loader:before,
  .loader:after {
    background: #ffffff;
    -webkit-animation: load1 1s infinite ease-in-out;
    animation: load1 1s infinite ease-in-out;
    width: 1em;
    height: 4em;
  }

  .loader {
    color: #ffffff;
    text-indent: -9999em;
    margin: 88px auto;
    position: relative;
    font-size: 11px;
    -webkit-transform: translateZ(0);
    -ms-transform: translateZ(0);
    transform: translateZ(0);
    -webkit-animation-delay: -0.16s;
    animation-delay: -0.16s;
  }

  .loader:before,
  .loader:after {
    position: absolute;
    top: 0;
    content: '';
  }

  .loader:before {
    left: -1.5em;
    -webkit-animation-delay: -0.32s;
    animation-delay: -0.32s;
  }

  .loader:after {
    left: 1.5em;
  }

  @-webkit-keyframes load1 {
    0%,
    80%,
    100% {
      box-shadow: 0 0;
      height: 4em;
    }
    40% {
      box-shadow: 0 -2em;
      height: 5em;
    }
  }

  @keyframes load1 {
    0%,
    80%,
    100% {
      box-shadow: 0 0;
      height: 4em;
    }
    40% {
      box-shadow: 0 -2em;
      height: 5em;
    }
  }
}
