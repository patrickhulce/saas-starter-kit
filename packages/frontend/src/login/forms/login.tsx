import * as React from 'react'

import {BasicTextField} from '../../components/basic-text-field'
import {Form, IFormData, IFormState, ISubmitOptions} from '../../components/form'
import {testIds} from '../../utils'

export async function login(email: string, password: string): Promise<void> {
  const authResponse = await fetch('/api/v1/oauth/token', {
    method: 'POST',
    body: JSON.stringify({username: email, password, grant_type: 'password'}),
    headers: {'content-type': 'application/json'},
    credentials: 'same-origin',
  })

  if (authResponse.status !== 200) {
    throw new Error('Unauthorized')
  }

  const userResponse = await fetch('/api/v1/users/me', {credentials: 'same-origin'})
  if (userResponse.status !== 200) {
    throw new Error('Invalid User')
  }

  const user = await userResponse.json()
  localStorage.setItem('loggedInUser', JSON.stringify(user))
  window.location.href = '/'
}

export class LoginForm extends Form {
  public state: IFormState = {}
  public testId: string = testIds.loginForm
  public submitUIOptions: ISubmitOptions = {label: 'Login'}

  protected async _handleSubmit(data: IFormData): Promise<void> {
    try {
      await login(data.email, data.password)
    } catch (err) {
      throw new Error('Invalid login')
    }
  }

  public renderInputUI(): JSX.Element {
    return (
      <React.Fragment>
        <BasicTextField name="email" type="email" autoFocus />
        <BasicTextField name="password" type="password" />
      </React.Fragment>
    )
  }
}
