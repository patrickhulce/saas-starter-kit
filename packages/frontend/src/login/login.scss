@import '../colors';
@import '../reset';

.header {
  margin-top: calc(50vh - 360px);
  margin-bottom: 40px;
  display: flex;
  flex-direction: column;
  align-items: center;
}

.logo {
  height: 200px;
  width: 200px;
  background: url('../assets/logo.svg') no-repeat;
}

h1.app-name {
  color: $accent-brand;
  font-size: 48px;
  font-weight: bold;
  text-shadow: 0 1px 5px $dark-gray;
  max-width: 100vw;
  overflow: hidden;
}

.account-forms {
  background: white;
  border-radius: 10px;
  width: 375px;
  min-height: 200px;
}

.newsletter {
  font-size: 0.9em;

  small {
    font-size: 0.6em;
    margin-top: -5px;
    display: block;
    color: $medium-gray;
  }
}

p.tos {
  text-align: center;
  margin-top: 20px;
}

@media (max-height: 800px) {
  .logo {
    display: none;
  }
}
