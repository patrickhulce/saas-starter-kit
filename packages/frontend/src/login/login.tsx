import * as React from 'react'
import {Form} from './form'
import Tabs, {Tab} from 'material-ui/Tabs'
import TextField from 'material-ui/TextField/TextField'
import Button from 'material-ui/Button/Button'

export async function login(email: string, password: string): Promise<void> {
  const response = await fetch('/api/v1/oauth/token', {
    method: 'POST',
    body: JSON.stringify({username: email, password, grant_type: 'password'}),
    headers: {'content-type': 'application/json'},
    credentials: 'same-origin',
  })

  if (response.status !== 200) {
    throw new Error('Unauthorized')
  }

  window.location.href = '/'
}

export class LoginForm extends Form {
  public async _handleSubmit(data: any): Promise<void> {
    try {
      await login(data.email, data.password)
    } catch (err) {
      this.setState({errorMessage: 'Invalid login'})
    }
  }

  public render(): JSX.Element {
    const errMsg = this.state.errorMessage ? <div>{this.state.errorMessage}</div> : undefined
    return (
      <form name='login' onSubmit={this.handleSubmit}>
        {errMsg}
        <TextField name='email' type='email' label='Email' autoFocus required />
        <TextField name='password' type='password' label='Password' required />
        <Button variant='raised' color='primary' type='submit'>
          Login
        </Button>
      </form>
    )
  }
}
