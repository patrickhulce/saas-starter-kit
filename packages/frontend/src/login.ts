/* tslint:disable */
import './login.scss'
import * as firebase from 'firebase'
import * as firebaseui from 'firebaseui'

const uiConfig = {
  tosUrl: 'THE_PRODUCT_WWW_URL/tos',
  signInOptions: [
    firebase.auth.EmailAuthProvider.PROVIDER_ID,
    firebase.auth.GoogleAuthProvider.PROVIDER_ID,
  ],
  callbacks: {
    uiShown() {
      document.getElementById('loader-container')!.style.display = 'none'
      document.getElementById('firebaseui-auth-container')!.style.display = 'block'
    },
    async signInSuccessWithAuthResult(authResult) {
      document.getElementById('loader-container')!.style.display = 'block'
      document.getElementById('firebaseui-auth-container')!.style.display = 'none'

      if (authResult.additionalUserInfo.isNewUser) {
        await fetch('/api/v1/accounts/register', {
          method: 'POST',
          body: JSON.stringify({
            name: `${authResult.user.displayName}'s Account`,
            email: authResult.user.email,
          }),
          headers: {'content-type': 'application/json'},
        })
      }

      const firebaseToken = await authResult.user.getIdToken()
      await fetch('/api/v1/oauth/tokens', {
        method: 'POST',
        body: JSON.stringify({username: '', password: firebaseToken}),
        headers: {'content-type': 'application/json'},
      })
    },
  },
}

const ui = new firebaseui.auth.AuthUI(firebase.auth())
ui.start('#firebaseui-auth-container', uiConfig)
