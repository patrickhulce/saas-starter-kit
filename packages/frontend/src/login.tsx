/* tslint:disable */
import * as styles from './login.scss'
import * as React from 'react'
import * as ReactDOM from 'react-dom'
import Tabs, {Tab} from 'material-ui/Tabs'
import TextField from 'material-ui/TextField/TextField'
import Button from 'material-ui/Button/Button'
import FormControlLabel from 'material-ui/Form/FormControlLabel'
import Checkbox from 'material-ui/Checkbox/Checkbox'
import FormHelperText from 'material-ui/Form/FormHelperText'

interface LoginPageState {
  selectedTab: number
  errorMessage?: string
}

async function createAccount(
  accountName: string,
  firstName: string,
  lastName: string,
  email: string,
  password: string,
): Promise<any> {
  const response = await fetch('/api/v1/accounts/register', {
    method: 'POST',
    body: JSON.stringify({
      account: {name: accountName},
      user: {firstName, lastName, email, password},
    }),
    headers: {'content-type': 'application/json'},
  })

  if (response.status !== 200) {
    const data = await response.json()
    throw new Error('Error creating account')
  }
}

async function login(email: string, password: string) {
  const response = await fetch('/api/v1/oauth/token', {
    method: 'POST',
    body: JSON.stringify({username: email, password, grant_type: 'password'}),
    headers: {'content-type': 'application/json'},
    credentials: 'same-origin',
  })

  if (response.status !== 200) {
    throw new Error('Unauthorized')
  }

  ;(window as any).location = '/'
}

class Form extends React.Component<{}, any> {
  constructor(props) {
    super(props)
    this.state = {}
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  handleSubmit(evt: any) {
    evt.preventDefault()
    const data = {}
    for (const [prop, val] of new FormData(evt.target).entries()) {
      data[prop] = val
    }

    this._handleSubmit(data)
  }

  _handleSubmit(data: any) {
    throw new Error('Unimplemented')
  }
}

class LoginForm extends Form {
  async _handleSubmit(data: any) {
    try {
      await login(data.email, data.password)
    } catch (err) {
      this.setState({errorMessage: 'Invalid login'})
    }
  }

  render() {
    const errMsg = this.state.errorMessage ? <div>{this.state.errorMessage}</div> : undefined
    return (
      <form name="login" onSubmit={this.handleSubmit}>
        {errMsg}
        <TextField name="email" type="email" label="Email" autoFocus required />
        <TextField name="password" type="password" label="Password" required />
        <Button variant="raised" color="primary" type="submit">
          Login
        </Button>
      </form>
    )
  }
}

class RegisterForm extends Form {
  async _handleSubmit(data: any) {
    if (data.cpassword !== data.password) {
      this.setState({errorMessage: 'Passwords did not match'})
      return
    }

    try {
      await createAccount(
        `${data.first} ${data.last}'s Account`,
        data.first,
        data.last,
        data.email,
        data.password,
      )
    } catch (err) {
      this.setState({errorMessage: 'Error creating account'})
    }
  }

  render() {
    const errMsg = this.state.errorMessage ? <div>{this.state.errorMessage}</div> : undefined
    return (
      <form name="register" onSubmit={this.handleSubmit}>
        {errMsg}
        <div>
          <TextField name="first" type="first" label="First Name" className="half-text" required />
          <TextField name="last" type="last" label="Last Name" className="half-text" required />
        </div>
        <TextField name="email" type="email" label="Email" required />
        <TextField name="password" type="password" label="Password" required />
        <TextField name="cpassword" type="password" label="Confirm Password" required />
        <FormControlLabel
          control={<Checkbox name="newsletter" value="newsletter" color="primary" />}
          label="Send me tips and updates (unsubscribe anytime)"
        />
        <FormHelperText className={styles.tos} >
          By registering, you agree to our <a href="TOS_URL">terms of service</a>
        </FormHelperText>
        <Button variant="raised" color="primary" type="submit">
          Register
        </Button>
      </form>
    )
  }
}

class LoginPage extends React.Component<{}, LoginPageState> {
  state = {selectedTab: 0}

  render() {
    const onChange = (evt, value) => this.setState({selectedTab: value})

    return [
      <h1 className="app-name">THE_PRODUCT_DISPLAY_NAME</h1>,
      <div className="account-forms">
        <Tabs
          value={this.state.selectedTab}
          onChange={onChange}
          indicatorColor="primary"
          textColor="primary"
          fullWidth
        >
          <Tab label="Login" />
          <Tab label="Create an Account" />
        </Tabs>
        {this.state.selectedTab === 0 && <LoginForm />}
        {this.state.selectedTab === 1 && <RegisterForm />}
      </div>,
    ]
  }
}

// HACK: let the CSS kick-in before rendering
setTimeout(() => {
  ReactDOM.render(<LoginPage />, document.getElementById('app-root'))
}, 0)

const anyModule = module as any
if (anyModule.hot) {
  anyModule.hot.accept(() => ReactDOM.render(<LoginPage />, document.getElementById('app-root')))
}
