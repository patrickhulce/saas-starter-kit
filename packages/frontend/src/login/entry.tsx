import '../typedefs' // tslint:disable-line
import * as React from 'react'
import * as ReactDOM from 'react-dom'
import Tabs, {Tab} from 'material-ui/Tabs'
import {LoginForm} from './login'
import {RegisterForm} from './register'

export interface ILoginPageState {
  selectedTab: number
  errorMessage?: string
}

export class LoginPage extends React.Component<{}, ILoginPageState> {
  public state: ILoginPageState = {selectedTab: 0}

  public render(): JSX.Element[] {
    const onChange = (evt, value) => this.setState({selectedTab: value})

    return [
      <h1 key="header" className="app-name">THE_PRODUCT_DISPLAY_NAME</h1>,
      <div key="form" className="account-forms">
        <Tabs
          value={this.state.selectedTab}
          onChange={onChange}
          indicatorColor="primary"
          textColor="primary"
          fullWidth
        >
          <Tab label="Login" />
          <Tab className="atmn-create-account" label="Create an Account" />
        </Tabs>
        {this.state.selectedTab === 0 && <LoginForm />}
        {this.state.selectedTab === 1 && <RegisterForm />}
      </div>,
    ]
  }
}

// HACK: let the CSS kick-in before rendering
setTimeout(() => {
  ReactDOM.render(<LoginPage />, document.getElementById('app-root'))
}, 0)

const anyModule = module as any
if (anyModule.hot) {
  anyModule.hot.accept(() => ReactDOM.render(<LoginPage />, document.getElementById('app-root')))
}
