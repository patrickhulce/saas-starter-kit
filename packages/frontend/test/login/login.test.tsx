import * as React from 'react'
import * as Enzyme from 'enzyme'
import {LoginForm} from '../../src/login/login'

describe('login/login.tsx', () => {
  it('should render form', () => {
    const form = Enzyme.shallow(<LoginForm />)
    expect(form).toMatchSnapshot()
  })
})
