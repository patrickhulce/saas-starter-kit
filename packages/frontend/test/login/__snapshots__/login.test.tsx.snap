// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login/login.tsx should render form 1`] = `
<form
  name="login"
  onSubmit={[Function]}
>
  <TextField
    autoFocus={true}
    label="Email"
    name="email"
    required={true}
    select={false}
    type="email"
  />
  <TextField
    label="Password"
    name="password"
    required={true}
    select={false}
    type="password"
  />
  <WithStyles(Button)
    color="primary"
    type="submit"
    variant="raised"
  >
    Login
  </WithStyles(Button)>
</form>
`;
